crates/matrix-sdk-appservice/src/lib.rs

```diff
@@ -627,6 +632,89 @@ impl AppService {
         webserver::warp_filter(self.clone())
     }
 
+    /// Receive an incoming [transaction], pushing the contained events to
+    /// active virtual clients.
+    ///
+    /// [transaction]: https://spec.matrix.org/v1.2/application-service-api/#put_matrixappv1transactionstxnid
+    pub async fn receive_transaction(
+        &self,
+        transaction: push_events::v1::IncomingRequest,
+    ) -> Result<()> {
+        let client = self.get_cached_client(None)?;
+
+        // Find membership events affecting members in our namespace, and update
+        // membership accordingly
+        for event in transaction.events.iter() {
+            let event = match event.deserialize() {
+                Ok(AnyRoomEvent::State(AnyStateEvent::RoomMember(event))) => event,
+                _ => continue,
+            };
+            if !self.user_id_is_in_namespace(event.state_key())? {
+                continue;
+            }
+            let localpart = event.state_key().localpart();
+            client
+                .store()
+                .set_custom_value(
+                    &[USER_MEMBER, event.room_id().as_bytes(), localpart.as_bytes(), b"."].concat(),
+                    event.membership().to_string().into_bytes(),
+                )
+                .await?;
+        }
+
+        #[derive(Debug, Deserialize)]
+        struct EventRoomId {
+            room_id: Option<OwnedRoomId>,
+        }
+
+        // Send a sync event to each active client
+        // TODO: What do we do with state events?
+        for virt_client in self.clients.iter() {
+            let user_id = match virt_client.user_id() {
+                Some(user_id) => user_id,
+                // Unauthenticated client. (should that be possible?)
+                None => continue,
+            };
+            let mut response = sync_events::v3::Response::new(transaction.txn_id.to_string());
+
+            for raw_event in &transaction.events {
```

Comment:
So the events app services receive are received in a flattened out structure. Clients on the other hand receive them in individual room buckets.

The important thing here is that each individual room  is in a bucket that denotes the room membership state of the current user. This bucketing is something the client expects.

Could we add a comment explaining this?

Also, this will modify and create personalized sync responses for each individual virtual user, has this been tried out in the real world, will this be slow? How many virtual users does an app-service usually handle?

It also seems that if we decide to do this we could, or rather should, parallelize this so each virtual client is handled in a task. Again, a TODO + issue would be enough.

Reply:
The main reasoning for taking the one sync response per virtual client approach was that it was the only way I could think of handler per-user membership properly without making deeper changes in `matrix-sdk`. It might be possible to reduce overhead by making each virtual user use the same store or something, but I'm assuming an approach in that direction would require deeper changes within the matrix-sdk crate to make that work properly

Reply:
Yeah, I assumed as much and that's understandable. Though would be nice to know if it has been tried out and if so under what load vs what an usual load for app-services is.

Reply:
Regarding parallelism, I'm thinking the `futures` crate's [`StreamExt::for_each_concurrent`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.for_each_concurrent) would be the easiest way of introducing concurrency, but I don't know if tokio is able to introduce maximum parallelism when using that, or if tasks are needed for that.

Reply:
Hm, making the parallelism work properly with lifetimes is turning out more complicated than expected, I'll leave it as a TODO so the focus can be on the basic behavior

Reply:
Sounds good.

Reply:
> Though would be nice to know if it has been tried out and if so under what load vs what an usual load for app-services is.

I'm not aware of any benchmarking or tests with large amounts of virtual users - so it's hard to say. However, I'd say that the solution outlined here already helps for the basic usage pattern and shows some of the needs for a solution that gets more deeply embedded into the sdk. Going by that I'd advocate for getting this merged so we can collect some real world experiences and then refactored before a crate release.